### ISSUE TITLE
Heap out of bounds read in sl_if_print()

### ISSUE THREAD

**Issue by fumfel** [the-tcpdump-group/tcpdump#564]
**Heap out of bounds read in sl_if_print()**

Tested on Git HEAD:  e2adb166ecaf59555e18787e86ac8c828efa0acc

To reproduce: `tcpdump -r oobr_sl_if_print.pcap`

![Crashing test case - please change extension to '.pcap'](https://cloud.githubusercontent.com/assets/5985287/20091904/7704b10a-a595-11e6-8fc7-c8cfff68d3c1.png)

ASAN output:

```
==12713==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60200000efc0 at pc 0x0000006738a6 bp 0x7fffc6a34960 sp 0x7fffc6a34958
READ of size 1 at 0x60200000efc0 thread T0
    #0 0x6738a5 in sl_if_print XYZ/tcpdump/./print-sl.c:78:10
    #1 0x501b12 in pretty_print_packet XYZ/tcpdump/./print.c:339:18
    #2 0x4eebeb in print_packet XYZ/tcpdump/./tcpdump.c:2492:2
    #3 0x7adaf7 in pcap_offline_read XYZ/libpcap/./savefile.c:527:4
    #4 0x6b6f4c in pcap_loop XYZ/libpcap/./pcap.c:890:8
    #5 0x4ea1db in main XYZ/tcpdump/./tcpdump.c:1996:12
    #6 0x7f2d4572282f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
    #7 0x419ff8 in _start (/usr/local/sbin/tcpdump+0x419ff8)

0x60200000efc0 is located 0 bytes to the right of 16-byte region [0x60200000efb0,0x60200000efc0)
allocated by thread T0 here:
    #0 0x4b880c in malloc /home/development/llvm/3.9.0/final/llvm.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:64:3
    #1 0x7af103 in pcap_check_header XYZ/libpcap/./sf-pcap.c:401:14
    #2 0x7acf22 in pcap_fopen_offline_with_tstamp_precision XYZ/libpcap/./savefile.c:400:7
    #3 0x7accaf in pcap_open_offline_with_tstamp_precision XYZ/libpcap/./savefile.c:307:6
    #4 0x7f2d4572282f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

SUMMARY: AddressSanitizer: heap-buffer-overflow XYZ/tcpdump/./print-sl.c:78:10 in sl_if_print
Shadow bytes around the buggy address:
  0x0c047fff9da0: fa fa 00 fa fa fa 00 fa fa fa 00 fa fa fa 07 fa
  0x0c047fff9db0: fa fa 00 fa fa fa 00 fa fa fa 05 fa fa fa 05 fa
  0x0c047fff9dc0: fa fa 07 fa fa fa 00 00 fa fa 00 00 fa fa 00 00
  0x0c047fff9dd0: fa fa 00 00 fa fa 00 00 fa fa 00 fa fa fa fd fd
  0x0c047fff9de0: fa fa fd fa fa fa fd fd fa fa fd fa fa fa fd fa
=>0x0c047fff9df0: fa fa fd fa fa fa 00 00[fa]fa 00 07 fa fa fd fd
  0x0c047fff9e00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c047fff9e10: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c047fff9e20: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c047fff9e30: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c047fff9e40: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==12713==ABORTING
```

**Comment by fxlb** [comment]
This one is already fixed in the security branch (will be merged in a few weeks).

**Comment by infrastation** [comment]
The original reporter for this vulnerability is Hanno Böck. It took a long time to publish the fix so the same issue was later independently discovered by other researchers.

Commit c4744fc fixes this and the fix is available in the release 4.9.0. Thanks to everybody for their reports, please send any future security reports to security@tcpdump.org.
